source/common/router/router.cc

```diff
@@ -724,8 +744,11 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::RequestTrailerMap& trail
 
 Http::FilterMetadataStatus Filter::decodeMetadata(Http::MetadataMap& metadata_map) {
   Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata_map);
-  ASSERT(upstream_requests_.size() == 1);
-  upstream_requests_.front()->encodeMetadata(std::move(metadata_map_ptr));
+  if (!upstream_requests_.empty()) {
+    // TODO: buffer this if there's no request? It doesn't look like the upstream request has
```

Comment:
router.h: TODO(soya3129): Save metadata for retry, redirect and shadowing case.  
Fine to double it here since it added confusion for you